### ISSUE TITLE
Final presentation brain storming

### ISSUE THREAD

**Issue by archive-cs-minhyuk** [HyunjoonCho/CS492I-IntroToDL-project#12]
### Problem definition (Motivation/Goals)

2020년 한국에서 시작된 '주식 열풍', 온 국민이 주식하는 시대
모두의 관심사: 어떻게 수익률을 높일 수 있는가?
가장 대표적인 방법, 가치 투자와 차트 분석!

가치 투자의 경우:  많은 사람들이 기업이 어떤 실적을 올리고 있는지, 앞으로 어떤 계획을 가지고 있는지, 기업의 재정 상태, 기업이 어떻게 운영되고 있는지 등을 생각함. 이를 효과적으로 반영할 수 있는 것이 뉴스라고 생각했음. 
-> 뉴스 기사를 이용해 미래 날짜의 주가가 어떻게 될지 예측해본다.

차트 분석의 경우: 많은 사람들이 5일선, 20일선 등 종가를 기준으로 한 '이동평균선'을 살핀다. 
-> 과거 일정 기간의 종가를 기준으로 미래 날짜의 주가가 어떻게 될지 예측해본다.

이렇게 구현된 두 모델을 이용해 해당 기업에 투자를 할지 고민하고 있는 사람들에게 다각도의 투자 조언을 해줄 수 있다.

### Method

**1) 기사를 바탕으로 한 주가 예측 모델(With OG dataset)**
 - 목표: 미래의 임의의 기사를 input으로 받아 그에 따른 주가변동폭을 예측하는 모델을 만들기. 주가변동폭은 '대폭 하락', '소폭 하락', '소폭 상승', '대폭 상승' 네 개의 class로 나누어 주어진 문제를 classification 문제로 풀어낸다.
 - 모델: BERT. 대충 다양한 분야에서 SOTA라는 내용... pretraining 과정은 훨씬 더 많은 계산 자원과 데이터를 요구하기 때문에 pretrained-BERT를 사용했다. 추후 영어 기사와 미국 주식으로도 확장하기 위해 다양한 언어에 사용 가능한 'multilingual BERT'를 사용함.
 - 데이터셋: 이 경우 우리의 목적에 부합하는 정확한 데이터셋이 없었기 때문에, 직접 데이터셋을 만들어야 했다. 데이터셋 내의 하나의 데이터는 (기사 제목들, 주가 변동폭)으로 구성된다. 우리의 문제가 classification problem이었기 때문에 4종류의 라벨을 가진 labelled data인 셈이다. 데이터를 모으기 위해 우리는 '웹 크롤링' 방법을 이용했고, 실제로 원하는 기업과 날짜를 주었을 때 그 날짜를 기준으로 하루 이전까지의 뉴스 기사 제목을 수집해 해당 날짜의 주가 변동폭과 연결시켜 데이터셋 내의 하나의 데이터를 만드는 웹 크롤러를 구현했다(이 부분이 설명이 어렵기 때문에 직관적으로 이해시킬 수 있는 예시 하나를 화면에 띄워야 할 듯). 이런 방법으로 우리는 코스피 시총 상위 50개 기업의 2018~2020년 기사를 이용한 데이터셋을 수집했고, 실제로 모델 training과 testing을 위해 5000개 가량의 데이터셋을 사용할 수 있었다.
  

> 여러 기사 제목이 있었을 것인데, 어떤 기준으로 기사 제목을 골랐는가: 그냥 해당 날짜의 관련 기업 뉴스 제목을 모두 이어 붙인 후 앞에서부터 64자 끊음. 

> 5000개의 dataset은 충분한 크기인가: pretrained BERT를 사용하여 fine-tuning하고 있고, 실제로 학습하는 부분은 BERT 위의 classification을 위한 additional layer이기 때문에, 충분하다고 생각한다. 실제로 뉴스 기사 분류 모델의 경우 2000개 미만의 데이터셋으로 높은 성능을 보여준다. 

> 크롤러 관련 더 설명해야 할 것: ‘구글 뉴스’ 이용, 단 구글 뉴스는 기업명을 이용해 불러왔을 때 기업명이 본문에 한 번이라도 포함된 뉴스를 모두 가져옴. 이러면 기업과 큰 관련이 없는 뉴스도 가져오기 때문에, 이를 해결하기 위해 제목에 기업명이 정확하게 들어간 뉴스들만 골라 제목을 모았음

- 결과: 전혀 정확도가 나오지 않는…(Training/Testing accuracy 포함한 캡쳐 필요할 듯)
 -> 이유? 기사와 주가변동폭 자체가 큰 관련이 없는 건 아닐까… 어떻게 하면 관련성을 높일 수 있을까? 우리가 흔히 투자를 위해 뉴스를 볼 때, 투자와 관련되었을 법한 뉴스만 골라서 보는 것을 착안
 

> 예를 들어 ‘삼성생명’이라는 기업에 관한 뉴스를 크롤링할 경우 기업 이름이 제목에 있다는 기준으로 필터링해도 주가와 전혀 관계없는 뉴스들이 나옴(삼성생명의 경우 배구/농구 팀이 있음). 이런 문제를 해결하기 위해 ‘뉴스 기사 종류 분류기’를 만들어 추가해보기로 결심 

**1-1) 기사와 주가변동폭의 관련도를 높이기 위한 ‘뉴스 기사 종류 분류기’ 모델**
 - 목표: 미래의 임의의 기사(제목)를 input으로 받아 그의 종류를 예측하는 모델을 만들기. 뉴스 기사 종류는 "정치","경제","사회", "생활/문화","세계","기술/IT", "연예", "스포츠" 여덟 개의 class로 나누어 주어진 기사를 하나의 세부 카테고리로 예측한다.
- 모델: BERT. 역시 위의 모델과 똑같이 pretrained-BERT를 사용했다. 여기서는 ‘KoBERT'를 사용함. 역시 BERT 위에 additional layer 얹어 분류하는 식.
- 데이터셋: 모 블로그(출처 밝히기)에서 가져온 8개의 카테고리, 각각 200개로 이루어진 dataset을 사용하여 훈련.
- 결과: test set에 대해서도 85% 이상의 높은 accuracy를 보이는 등, 높은 성능의 분류기 완성. 이제 완성된 분류기를 이용해 더욱 ‘정제된’ dataset을 수집한다. 위의 크롤러에 모델을 더하는 방법인데, 원하는 기업과 날짜를 입력해 기사 제목을 가져오는 과정까지는 똑같다. 그 기사 제목을 분류기 모델에 넣어 결과가 ‘경제’, ‘기술/IT’일 경우에만 제목을 모으고, 나머지는 버림. 이렇게 해서 (경제, IT 뉴스만 모은 제목들, 해당 날짜의 주가 변동폭)의 dataset이 완성되었다!

**1-2) 기사를 바탕으로 한 주가 예측 모델(With New dataset)**

(input부터 output까지 전체적인 flow를 나타내는 그림이 하나 필요할 듯)
- 결과: 25%에서 크게 벗어나지 못하던 이전 모델과 달리, 40% 정도의 최고 test accuracy를 기록하는 등의 발전을 보임(결과 캡쳐 필요). Overfitting과 early stopping에 대해 이야기해야 할 듯(40epoch까지 training 해본 결과, 15 epoch 즈음이 test accuracy가 가장 높은 시점이었다)

**2) 과거 일정 기간의 종가를 바탕으로 한 주가 예측 모델**
 - 목표: 특정 날짜를 input으로 받아 그 이전 20일의 주가를 보고 해당 날짜의 주가 변동폭을 예측하는 모델을 만들기. 실제 가격을 예측하는 regression 모델이지만, 최종 결과는 위의 기사 output과 맞추기 위해 주가 + 주가변동폭으로 알려준다.
 - 모델: LSTM. 순서가 중요한 feature의 요소일 때 흔히 RNN을 적용, 그 중에서도 LSTM이 가장 높은 성능! 주가와 같은 시계열 데이터셋에서 효과적인 모델.
 - 데이터셋: 이 경우에는 어렵지 않게 직접 데이터셋을 만들 수 있음. python의 ‘finance-data-reader’라는 library를 사용하면 원하는 기업의 과거 주가(시가, 고가, 수급, 종가 등 포함)를 20년 어치 넘게 가져올 수 있는데, 해당 데이터를 20개씩 묶어 마지막 날짜의 종가를 label로 사용함으로써 training/testing을 진행할 수 있다.
- 결과: (stock price 그래프와 training loss 그래프 스크린샷 첨부, training RMSE와 testing RMSE도 함께 첨부하면 좋을 듯) training에서 좋은 결과. Testing 역시 마찬가지. 비록 20일 전 데이터를 가져오는 과정을 반복했기 때문에 보이는 그래프만큼 미래 예측이 정확하다는 보장은 없지만, 어쨌든 차트를 읽을 정도의 역할은 할 수 있음. 그 이후 실제 기업/날짜 설정해 예측하는 것도 보여주고 말하면 됨.

**3) 모든 것을 합친 모델 (Should I Invest?)**
- 모델: (모두 합친 모델, input부터 output까지 전체적인 flow가 나타나도록)
- 결과: (원하는 기업/날짜 설정 + 결과 보여주는 것 스크린샷) 혹은 나중에 시각화 구현을 한다면 이 부분에 추가

### Conclusion (Limitations/Future Works)

**Limitations**
-	기사 기반 주가 예측의 경우, 낮은 정확도(Test accuracy 최대 40퍼에 그침)
  : 주가와 관련 있는 기사를 모으려는 노력을 조금 더 했을 때 정확도가 높아질 수 있을 것으로 기대됨.
     혹은 더 많은 크롤링을 통해 데이터셋 자체의 크기를 늘리는 것도 모델이 경향성을 습득하는 데에 도움을 줄 수 있다고 생각

-	To be added
